tools/api-markdown-documenter/src/renderers/DocumentWriter.ts

```diff
@@ -2,20 +2,311 @@
+ * Can be instantiated via {@link (DocumentWriter:namespace).create}.
  *
+ * @remarks Note that the indentation is inserted at the last possible opportunity.
+ *
+ * @example
+ *
+ * For example, this code...
+ *
+ * ```ts
+ *   writer.write('begin\n');
+ *   writer.increaseIndent();
+ *   writer.write('one\ntwo\n');
+ *   writer.decreaseIndent();
+ *   writer.increaseIndent();
+ *   writer.decreaseIndent();
+ *   writer.write('end');
+ * ```
+ *
+ * ...would produce this output:
+ *
+ * ```
+ *   begin
+ *     one
+ *     two
+ *   end
+ * ```
+ *
+ * @public
+ */
+export interface DocumentWriter {
+	/**
+	 * Retrieves the output that was built so far.
+	 */
+	getText(): string;
+
+	/**
+	 * Increases the indentation.
+	 *
+	 * @param indentPrefix - The character(s) to append to the line indentation.
+	 * For example, the prefix could be "// " to indent and comment simultaneously.
+	 * Default: {@link (DocumentWriter:namespace).defaultIndentPrefix}.
+	 *
+	 * @remarks Each call to `increaseIndent` must be followed by a corresponding call to {@link (DocumentWriter:interface).decreaseIndent}.
+	 */
+	increaseIndent(indentPrefix?: string): void;
+
+	/**
+	 * Decreases the indentation, reverting the effect of the corresponding call
+	 * to {@link (DocumentWriter:interface).increaseIndent}.
+	 */
+	decreaseIndent(): void;
+
+	/**
+	 * Adds a newline if the file pointer is not already at the start of the line (or start of the stream).
+	 */
+	ensureNewLine(): void;
+
+	/**
+	 * Adds up to two newlines to ensure that there is a blank line above the current line.
+	 */
+	ensureSkippedLine(): void;
+
+	/**
+	 * Writes some text to the internal string buffer, applying indentation according to the current indentation level.
+	 * @remarks If the string contains multiple newlines, each line will be indented separately.
+	 */
+	write(message: string): void;
+
+	/**
+	 * A shorthand for writing an optional message, followed by a newline.
+	 * @remarks Indentation is applied following the semantics of {@link (DocumentWriter:interface).write}.
+	 */
+	writeLine(message?: string): void;
+
+	/**
+	 * Returns the last character that was written, or and empty string if no characters have been written yet.
+	 */
+	peekLastCharacter(): string;
+
+	/**
+	 * Returns the second to last character that was written, or an empty string if fewer than two characters
+	 * have been written yet.
+	 */
+	peekSecondLastCharacter(): string;
+}
+
+/**
+ * {@inheritDoc (DocumentWriter:interface)}
  * @public
  */
-export function createDocumentWriter(): DocumentWriter {
-	return new DocumentWriter(new StringBuilder());
+// eslint-disable-next-line @typescript-eslint/no-namespace
+export namespace DocumentWriter {
+	/**
+	 * The text characters used to create one level of indentation.
+	 * Default: 2 spaces.
+	 */
+	export const defaultIndentPrefix: string = "  "; // TODO: consider using a tab instead.
```

Comment:
Nit: not sure I can think of a scenario where exporting this would be super useful. Not blocking, just trying to apply the rule of thumb of not exporting unless we actually need it.

Reply:
Hmm. I don't know why I exported this. It might have been with the assumption that it would be used in testing. Good catch!

Reply:
```suggestion
	const defaultIndentPrefix: string = "  "; // TODO: consider using a tab instead.
```